[jx124] iP

jx124 opened the pull request:
# EnPassant
> "One must imagine Sisyphus happy." - Albert Camus, _The Myth of Sisyphus_.

With EnPassant, you don't have to think anymore. You just do. EnPassant is
* text-based
* easy to learn and use
* ~super~ _**blazingly**_ fast

To start, all you need to do is
1. download EnPassant from [here](https://github.com/jx124/ip/releases/tag/A-Jar)
2. double-click on it
3. add your tasks
4. let EnPassant manage it for you!

How is this incredible program free??! 😱😱

Features:
- [x] managing to-dos
- [x] managing deadlines
- [ ] reminders (coming soon)


If you are a Java programmer, here's the `main` method for you to play around in:
```java
public static void main(String[] args) {
    new Duke("./data/duke.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jx124 committed:
Rename bot

jx124 committed:
Add echo and exit functionality

jx124 committed:
Implement add and list functionality

jx124 committed:
Add Task class

jx124 committed:
Implement mark as done functionality

jx124 committed:
Add list message

jx124 committed:
Add Todo, Deadline, Event subclasses

jx124 committed:
Refactor main function to use regex and switch cases, and to include new Tasks

jx124 committed:
Add test input and expected output

jx124 committed:
Add DukeException class

jx124 committed:
Add exceptions handling

jx124 committed:
Add delete Task feature

jx124 committed:
Update test files

jx124 committed:
Add javadoc documentation

jx124 committed:
Change class to abstract class and add getSaveString abstract method

jx124 committed:
Implement getSaveString method

jx124 committed:
Add save file parsing and task saving

jx124 committed:
Add data directory to gitignore

jx124 committed:
Merge branch 'branch-Level-7'

jx124 committed:
Modify Deadline task to store LocalDateTime

jx124 committed:
Modify Duke to parse and save LocalDateTime for Deadline tasks

jx124 committed:
Merge branch 'branch-Level-8'

jx124 committed:
Remove need for separate directory path by getting file path parent

jx124 committed:
Create Ui class

jx124 committed:
Create and implement Storage class

jx124 committed:
Incorporate Storage class into Duke

jx124 committed:
Implement Ui class

jx124 committed:
Incorporate Ui class

jx124 committed:
Create and move all messages to new Messages class

jx124 committed:
Create new abstract Command class

jx124 committed:
Implement ExitCommand

jx124 committed:
Implement ListCommand

jx124 committed:
Switch to using messages from the Messages class

jx124 committed:
Move storage message to Messages class

jx124 committed:
Modify execute method to throw DukeException

jx124 committed:
Implement MarkCommand

jx124 committed:
Implement UnmarkCommand

jx124 committed:
Implement TodoCommand

jx124 committed:
Rename TodoCommand to TaskCommand: all Tasks can use the same command

jx124 committed:
Implement DeleteCommand

jx124 committed:
Add invalid index check

jx124 committed:
Implement Parser class

jx124 committed:
Change field and parse method to be static, remove constructor

jx124 committed:
Implement TaskList

jx124 committed:
Refactor to use TaskList instead of ArrayList

jx124 committed:
Merge branch 'branch-A-MoreOOP'

jx124 committed:
Refactor code using packages

jx124 committed:
Merge branch 'branch-A-Packages'

jx124 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jx124 committed:
Move bounds checking to TaskList methods

jx124 committed:
Add tests for TaskList

jx124 committed:
Add tests for Parser

jx124 committed:
Merge branch 'branch-A-JUnit'

jx124 committed:
Modify main class location

jx124 committed:
Merge branch 'branch-A-Jar'

jx124 committed:
Add JavaDoc for classes and methods in the main duke package

jx124 committed:
Add package name prefixes

jx124 committed:
Rename TaskCommand class

jx124 committed:
Add JavaDoc to classes and methods in duke.command package

jx124 committed:
Add newline after JavaDoc descriptions

jx124 committed:
Reformat code

jx124 committed:
Add new messages for the find command

jx124 committed:
Implement findTasks method

jx124 committed:
Implement FindCommand

jx124 committed:
Incorporate find command into parser

jx124 committed:
Merge branch 'branch-A-JavaDoc'

jx124 committed:
Resolve marge conflicts

jx124 committed:
Resolve merge conflicts

jx124 committed:
Rename messages to have MESSAGE as a prefix

jx124 committed:
Expand wildcard imports

AlyssaPng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+import duke.command.UnmarkCommand;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
```
```suggestions
import duke.command.AddTaskCommand;
import duke.command.Command;
import duke.command.DeleteCommand;
import duke.command.ExitCommand;
import duke.command.FindCommand;
import duke.command.ListCommand;
import duke.command.MarkCommand;
import duke.command.UnmarkCommand;

import duke.task.Deadline;
import duke.task.Event;
import duke.task.Todo;
```

AlyssaPng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+public class Duke {
+    private final Storage storage;
+    private final Ui ui;
+    private TaskList tasks;
```
Good Job on using plural form for tasks!

AlyssaPng commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return A formatted string with the relevant information for the save file.
+     */
+    public abstract String getSaveString();
```
I like how you used a abstract method here.

AlyssaPng commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+public class Ui {
+    public static final String LINE = "____________________________________________________________\n";
+    private final Scanner in;
+    private final PrintStream out;
```
is this naming sufficient for you to understand what the fields are for?

AlyssaPng reviewed (commented):
Looks Good :)
Comments are for Coding Standards

jx124 committed:
Add basic JavaFX template

jx124 committed:
Change Ui class methods to return string instead of printing

jx124 committed:
Change Command execute method to return string response

jx124 committed:
Remove UI line element from exception messages

jx124 committed:
Implement proper user input response

jx124 committed:
Merge branch 'branch-Level-10'

maj0-0 commented on src/main/java/duke/DialogBox.java:
Overall, the DialogBox class is well-written, follows good coding practices, and is easy to understand. It effectively encapsulates the UI elements for displaying user and Duke dialogs in a JavaFX application!

maj0-0 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param filePath The location of the save file to be created at and read from.
+     */
+    public Duke(String filePath) {
```
I like that the constructor and the getResponse method are appropriately sized and do not exhibit deep nesting!

maj0-0 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,9 @@
+package duke;
+
+import javafx.application.Application;
+
```
It would be helpful if you could add Javadocs to explain the purpose of the Launcher class!

maj0-0 commented on src/main/java/duke/Main.java:
In my opinion, the Main class is well-structured for launching the Duke application as a JavaFX application. It adheres to good coding practices, enhancing its readability and maintainability. The code is concise, clear, and effectively serves its purpose!

maj0-0 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,157 @@
+     */
+    public ArrayList<Task> loadTasks() throws DukeException {
+        File file = this.getOrCreateFile();
+        String regexPattern = "([TDE])\\s\\|\\s"    // match type
```
Consider declaring constants for recurring values like the regex pattern to enhance maintainability further.

maj0-0 reviewed (commented):
Overall, your code for the Duke application appears to be well-structured and follows good coding practices. It appears to be well-prepared to handle various user interactions for task management!

jx124 committed:
Add not null assertions to Command subclasses

The method execute in these classes do not check if the arguments passed in are null.
Adding these assertions could help to prevent null pointer exceptions or
improper use of the methods.

jx124 committed:
Add javadoc for Launcher class

jx124 committed:
Refactor Parser parse method

The parse method contains complicated boolean expressions.

Splitting the expression up into simpler portions make it easier to maintain.

Let's split the expression into two parts, isArgumentEmpty to check if the
relevant arguments are filled in, and isKeywordPresent to check if the correct
keywords are used.

jx124 committed:
Storage class: extract regex pattern from loadTasks

The loadTasks method initializes a new string containing the regex pattern every time it
is called.

Since it is a constant, extracting it our of the method can make it more performant and
readable.

Let's replace it with a private static final String instead.

jx124 committed:
Replace magic string literal

jx124 committed:
Merge pull request #1 from jx124/branch-A-Assertions

Add not null assertions to Command subclasses

jx124 committed:
Merge pull request #2 from jx124/branch-A-CodeQuality

Improve code quality

jx124 committed:
Implement Comparator class for Tasks

jx124 committed:
Add sorting functionality

jx124 committed:
Add javadoc to TaskList methods

jx124 committed:
Merge branch 'branch-C-Sort'

jx124 committed:
Refactor Parser class parse method

The parse method is too long and does not adhere to SLAP.

Extracting the bodies of the case statements into individual
functions would fix this issue and make the code more readable.

Let's create a method to parse the regex Matcher into a Command
for every branch of the switch case.

jx124 committed:
Refactor Storage class loadTasks method

The loadTasks method is too long and does not adhere to SLAP.

Extracting the bodies of the case statements into individual
functions would fix this issue and make the code more readable.

Let's create a method to load each type of Task into the TaskList
for every branch of the switch case.

jx124 committed:
Rephrase Javadoc overview statements to match coding standard

jx124 committed:
Merge branch 'branch-CodeQualityFeedback'

jx124 committed:
Make main window resizable

jx124 committed:
Make text label fully expand

jx124 committed:
Add padding to label

jx124 committed:
Merge branch 'branch-A-BetterGui'

jx124 committed:
Add name to title bar

jx124 committed:
Add product screenshot

jx124 committed:
Update user guide
